design/machine-remediation-request.md

```diff
@@ -0,0 +1,142 @@
+<!--
+ This work is licensed under a Creative Commons Attribution 3.0
+ Unported License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# machine-remediation-request
+
+## Status
+
+in-progress
+
+## Table of Contents
+
+<!--ts-->
+* [machine-remediation-request](#machine-remediation-request)
+  * [Status](#status)
+  * [Table of Contents](#table-of-contents)
+  * [Summary](#summary)
+  * [Motivation](#motivation)
+    * [Goals](#goals)
+    * [Non-Goals](#non-goals)
+  * [Proposal](#proposal)
+    * [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+      * [MachineRemediationRequest spec API](#machineremediationrequest-spec-api)
+      * [MachineRemediationRequest status API](#machineremediationrequest-status-api)
+    * [Risks and Mitigations](#risks-and-mitigations)
+  * [Design Details](#design-details)
+    * [Work Items](#work-items)
+    * [Dependencies](#dependencies)
+    * [Test Plan](#test-plan)
+    * [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+    * [Version Skew Strategy](#version-skew-strategy)
+  * [Alternatives](#alternatives)
+  * [References](#references)
+
+<!-- Added by: dhellmann, at: Wed Jun 26 16:08:57 EDT 2019 -->
+
+<!--te-->
+
+## Summary
+
+This document explains the implementation of the Machine Remediation Request controller and
+how it will work together with Machine Health Check controller to restore unhealthy machines.
+
+## Motivation
+
+To provide a mechanism to apply remediation strategy on an unhealthy node, that will be different by type and by a provider.
+
+### Goals
+
+1. To provide the interface that will make possible to apply different remediation strategy for each cloud or bare metal providers.
+2. To provide details regarding the progress of the remediation operation.
+
+### Non-Goals
+
+1. To provide a remediation strategy for each cloud or bare metal providers.
+2. To provide the integration with the **MachineHealthCheck** controller.
+
+## Proposal
+
+### Implementation Details/Notes/Constraints
```

Comment:
How will the controller implement reboot? Particularly how will it do it for different providers?

Reply:
IIUC, the only provider we will create a controller for initially is metal3.
Exactly how each controller "fixes" the affected machine is intended to be opaque to MachineHealthCheck.

This allows the metal3 flavor of the MachineRemediationRequest controller to look up and use Spec.Online and Status.PoweredOn from the BareMetalHost to implement reboot.

Reply:
Actually more than the above, the current intention for 4.2 is to only create the CR if the bare metal host annotation is found on the Machine object.  We would then expand to other providers as appropriate in later versions.

Reply:
So each provider will need its own machine-remediation-controller implementation?

Reply:
yep, it was the idea, for the short term we will implement only bare metal remediator, in general remediation can apply additional actions depends on a provider and the infrastructure configure on the node.